src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,477 @@
+  return cov_poses_.size() != 0;
+}
+
+bool BundleAdjustmentCovarianceEstimatorBase::HasValidFullCovariance() const {
+  return cov_variables_.size() != 0;
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetPoseCovariance()
+    const {
+  THROW_CHECK(HasValidPoseCovariance());
+  return cov_poses_;
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetPoseCovariance(
+    image_t image_id) const {
+  THROW_CHECK(HasValidPoseCovariance());
+  THROW_CHECK(HasPose(image_id));
+  int index = GetPoseIndex(image_id);
+  int num_params_pose = GetPoseTangentSize(image_id);
+  return cov_poses_.block(index, index, num_params_pose, num_params_pose);
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetPoseCovariance(
+    const std::vector<image_t>& image_ids) const {
+  THROW_CHECK(HasValidPoseCovariance());
+  std::vector<int> indices;
+  for (const auto& image_id : image_ids) {
+    THROW_CHECK(HasPose(image_id));
+    int index = GetPoseIndex(image_id);
+    int num_params_pose = GetPoseTangentSize(image_id);
+    for (int i = 0; i < num_params_pose; ++i) {
+      indices.push_back(index + i);
+    }
+  }
+#if EIGEN_VERSION_AT_LEAST(3, 4, 0)
+  Eigen::MatrixXd output = cov_poses_(indices, indices);
+#else
+  size_t n_indices = indices.size();
+  Eigen::MatrixXd output(n_indices, n_indices);
+  for (size_t i = 0; i < n_indices; ++i) {
+    for (size_t j = 0; j < n_indices; ++j) {
+      output(i, j) = cov_poses_(indices[i], indices[j]);
+    }
+  }
+#endif
+  return output;
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetPoseCovariance(
+    image_t image_id1, image_t image_id2) const {
+  THROW_CHECK(HasValidPoseCovariance());
+  THROW_CHECK(HasPose(image_id1));
+  THROW_CHECK(HasPose(image_id2));
+  int index1 = GetPoseIndex(image_id1);
+  int num_params_pose1 = GetPoseTangentSize(image_id1);
+  int index2 = GetPoseIndex(image_id2);
+  int num_params_pose2 = GetPoseTangentSize(image_id2);
+  return cov_poses_.block(index1, index2, num_params_pose1, num_params_pose2);
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetCovariance(
+    double* params) const {
+  THROW_CHECK(HasValidFullCovariance());
+  THROW_CHECK(HasBlock(params));
+  int index = GetBlockIndex(params);
+  int num_params = GetBlockTangentSize(params);
+  return cov_variables_.block(index, index, num_params, num_params);
+}
+
+Eigen::MatrixXd BundleAdjustmentCovarianceEstimatorBase::GetCovariance(
+    const std::vector<double*>& blocks) const {
+  THROW_CHECK(HasValidFullCovariance());
+  std::vector<int> indices;
+  for (const auto& block : blocks) {
+    THROW_CHECK(HasBlock(block));
+    int index = GetBlockIndex(block);
+    int num_params_pose = GetBlockTangentSize(block);
+    for (int i = 0; i < num_params_pose; ++i) {
+      indices.push_back(index + i);
+    }
+  }
+#if EIGEN_VERSION_AT_LEAST(3, 4, 0)
+  Eigen::MatrixXd output = cov_variables_(indices, indices);
```

Comment:
Is this faster than the manual loop below? If not, then it might not be worth the ifdef?